src/core/defs.jl

```diff
@@ -187,6 +187,24 @@ dim_value_dict(md::ModelDef) = Dict([name => collect(values(dim)) for (name, dim
 Base.haskey(md::ModelDef, name::Symbol) = haskey(md.dimensions, name)
 
 isuniform(md::ModelDef) = md.is_uniform
+
+
+# Helper function invoked when the user resets the time dimension with set_dimension!
+# This function calls set_run_period! on each component definition to reset the first and last values.
+function reset_run_periods!(md, first, last)
+
+    for comp_def in values(md.comp_defs)
+        if comp_def.first < first 
+            warn("Resetting $(comp_def.name) component's first timestep to $first")
+        end 
+        if comp_def.last > last 
+            warn("Resetting $(comp_def.name) component's last timestep to $last")
+        end
+        set_run_period!(comp_def, max(comp_def.first, first), min(comp_def.last, last))
```

Comment:
This is admittedly minor, but calling set_run_period! here happens even when no change is necessary (even though we checked this immediately above) and we call `max` and `min` to recover information we also had already. And it's in a loop over all components. You could economize several ways. Here's one:
```
function reset_run_periods!(md, first, last)
    for comp_def in values(md.comp_defs)
        changed = false
        if comp_def.first < first 
            warn("Resetting $(comp_def.name) component's first timestep to $first")
            changed = true
        else
            first = comp_def.first
        end 

        if comp_def.last > last 
            warn("Resetting $(comp_def.name) component's last timestep to $last")
            changed = true
        else
            last = comp_def.last
        end

        if changed
            set_run_period!(comp_def, first, last)
        end
    end

    nothing
end
```